Homework_11/hw11_internet.py

```diff
@@ -0,0 +1,152 @@
+
+
+# Task1
+response = requests.get("https://www.imdb.com/chart/top/?ref_=nv_mp_mv250")
+soup = BeautifulSoup(response.content, "lxml")
+
+films_dict = {}
+# tags list with title name, year, and director
+rating_list = soup.find_all("td", attrs={"class": "titleColumn"})
+
+# extracting title names
+titles = [title.find('a').get_text() for title in rating_list]
+films_dict.update({"name": titles})
+
+# extracting ranks
+ranks = [int(i.a.previous_element.strip()[:-1]) for i in rating_list]
+films_dict.update({"rank": ranks})
+
+# extracting years
+years_pattern = re.compile(r'\d{4}')
+years = [re.search(years_pattern, title.find('span').get_text())[0] for title in rating_list]
+films_dict.update({"year": years})
+
+# extracting ratings and n_reviews
+rate_reviews_list = soup.find_all("td", attrs={"class": "ratingColumn imdbRating"})
+rate_reviews_pattern = re.compile(r'(\d+([,\.]{1}\d+)+)')
+films_dict.update({"rating": [], "n_reviews": []})
+for i in rate_reviews_list:
+    (mark, _), (n_votes, _) = re.findall(rate_reviews_pattern, i.strong.attrs["title"])
+    films_dict["rating"].append(float(mark))
+    films_dict["n_reviews"].append(int(n_votes.replace(",", "")))
+
+# extracting directors    
+directors = [direc.a.attrs["title"].split(" (")[0] for direc in rating_list]
+films_dict.update({"directors": directors})
+films_df = pd.DataFrame(films_dict)
+
+# Subtask1
+top_reviewed = films_df[["name", "n_reviews"]].sort_values(by="n_reviews", ascending=False).head(4)
+top_reviewed
+
+# Subtask2
+top_years = films_df.groupby("year")["rating"].mean().sort_values(ascending=False).head(4)
+top_years
+
+# Subtask3
+directors_count  = films_df["directors"].value_counts()
+directors_count = directors_count[directors_count > 2]
+
+fig, ax = plt.subplots(figsize=(10, 6), dpi=300)
+sns.barplot(x=directors_count.index, y=directors_count.values, palette="twilight", ax=ax)
+ax.set_title("Films number for directors from Top 250 IMDb", fontsize=14, fontweight='bold')
+ax.set_xlabel("director", fontweight='bold')
+ax.set_xticklabels(directors_count.index, rotation=45, ha="right")
+ax.set_ylabel("count", fontweight='bold');
+
+# Subtask4
+films_df.groupby("directors")["n_reviews"].sum().sort_values(ascending=False).head(4)
+
+# Subtask5
+print(films_df.shape)
+films_df.head()
+
+# Task2
+load_dotenv()
+TG_API_TOKEN = os.getenv("TG_API_TOKEN")
+chat_id = '133649047'
+
+def telegram_logger(chat_id):
+    '''A decorator for calculating function exacution time, 
+    creating .log file, and sending it to telegram bot
+    
+    Parameters:
+        chat_id (str): chat id for sending messages by bot
+    '''
+    
+    
+    def wrapper(func):
+        def inner_func(*args, **kwargs):
+            url = f'https://api.telegram.org/bot{TG_API_TOKEN}/'
+            old_stdout, old_stderr = sys.stdout, sys.stderr
+            text_buffer = io.StringIO()
+            sys.stdout = sys.stderr = text_buffer
+            
+            try:
+                start = time.time()
```

Comment:
При вычитании двух `datetime` автоматически получается `timedelta`, очень удобно
```suggestion
                start = datetime.now()
```